python/pyarrow/_compute.pyx

```diff
@@ -2251,3 +2325,158 @@ cdef CExpression _bind(Expression filter, Schema schema) except *:
+        number of elements equal to the num_args (which is the miniumu
+        required arguments).
+    out_type : DataType
+        Output type of the function.
+    function : callable
+        User-defined-function
+        function includes arguments equal to the number
+        of input_types defined. The return type of the 
+        function is of the type defined as output_type. 
+        The output should be an Array, Scalar, ChunkedArray,
+        Table, or RecordBatch based on the out_type.
+
+    Example
+    -------
+
+    >>> from pyarrow import compute as pc
+    >>> from pyarrow.compute import register_scalar_function
+    >>> from pyarrow.compute import InputType
+    >>> 
+    >>> func_doc = {}
+    >>> func_doc["summary"] = "simple udf"
+    >>> func_doc["description"] = "add a constant to a scalar"
+    >>> func_doc["arg_names"] = ["x"]
+    >>> 
+    >>> def add_constant(array):
+    ...     return pc.call_function("add", [array, 1])
+    ... 
+    >>> 
+    >>> func_name = "py_add_func"
+    >>> arity = 1
+    >>> in_types = [InputType.array(pa.int64())]
+    >>> out_type = pa.int64()
+    >>> register_function(func_name, arity, func_doc,
+    ...                   in_types, out_type, add_constant)
+    >>> 
+    >>> func = pc.get_function(func_name)
+    >>> func.name
+    'py_add_func'
+    >>> ans = pc.call_function(func_name, [pa.array([20])])
+    >>> ans
+    <pyarrow.lib.Int64Array object at 0x10c22e700>
+    [
+    21
+    ]
+    """
+    cdef:
+        c_string c_func_name
+        CArity c_arity
+        CFunctionDoc c_func_doc
+        CInputType in_tmp
+        vector[CInputType] c_in_types
+        PyObject* c_function
+        shared_ptr[CDataType] c_type
+        COutputType* c_out_type
+        CScalarUdfBuilder* c_sc_builder
+        CStatus st
+        CScalarUdfOptions* c_options
+
+    c_func_name = tobytes(func_name)
+
+    if callable(function):
+        c_function = <PyObject*>function
+    else:
+        raise ValueError("Object must be a callable")
+
+    func_spec = inspect.getfullargspec(function)
+    if func_spec.varargs:
+        if num_args <= 0:
+            raise ValueError("number of arguments must be >= 0")
+        c_arity = CArity.VarArgs(num_args)
+    else:
+        if num_args <= 0:
+            raise ValueError("number of arguments must be >= 0")
+        if num_args == 0:
+            c_arity = CArity.Nullary()
+        elif num_args == 1:
+            c_arity = CArity.Unary()
+        elif num_args == 2:
+            c_arity = CArity.Binary()
+        elif num_args == 3:
+            c_arity = CArity.Ternary()
+
+    c_func_doc = _make_function_doc(function_doc)
+
+    if in_types and isinstance(in_types, list):
+        for in_type in in_types:
+            in_tmp = (<InputType> in_type).input_type
+            c_in_types.push_back(in_tmp)
+    else:
+        raise ValueError("input types must be of type InputType")
```

Comment:
Do we expect user to pass a Generator? Any specific use case? I assumed a List would be enough for this. I might have missed it.

Reply:
@lidavidm do you have a usecase for this?

Reply:
Python is not so strictly typed. A user might expect to pass in a tuple at the very least.

However, the code has changed since then and `dict` is the only reasonable type. In general, however, `isinstance` is likely a code smell in Python. Just use the value and let Python raise an exception if it does not "work" (duck typing). If you are going to raise an exception yourself, at least make sure it is TypeError and make sure your error message is actually correct.